modules/DrugSynonymDataToNeo4j.py

```diff
@@ -31,55 +32,105 @@ def cypher_template_filler(cypher_template:str,data:dict) -> str:
-
+        self._driver = GraphDatabase.driver(uri, auth=(user, password), encrypted=encrypted)	        
+        self.study_triald_and_neo4j_id_pairs:dict = {}
+        self.drug_or_synonym_name_and_neo4j_id_pairs:dict = {}
+        self.url_and_neo4j_id_pairs:dict = {}
+
+    def reset_id_store(self):
+        self.study_triald_and_neo4j_id_pairs = {}
+        self.drug_or_synonym_name_and_neo4j_id_pairs = {}
+        self.url_and_neo4j_id_pairs = {}
+    
     def close(self):
         self._driver.close()
+    
+    @staticmethod
+    def _merge_node(tx, node_type, properties:Optional[dict] = None):
 
-    def upload_studies(self,studies:DataFrame):
+
+
+        data:dict = {
+            "node_type":node_type,
+            "properties":dict_to_property_str(properties)
+        }
+        base_cypher = """
+            MERGE (n:{node_type} {properties})
+            RETURN id(n)
+        """
+
+        result = tx.run(cypher_template_filler(base_cypher,data))
+        return result.single()[0]
+    
+    @staticmethod
+    def _merge_edge(tx, from_id, to_id, edge_type, properties:Optional[dict] = None, direction = ">"):
+        if not direction in [">",""]:
+            raise ValueError
+
+        data:dict = {
+            "from_id":int(from_id),
+            "to_id":int(to_id),
+            "edge_type":edge_type,
+            "direction":direction,
+            "properties":dict_to_property_str(properties) 
+        }
+        base_cypher = """
+            MATCH (from)
+            WHERE ID(from) = {from_id}
+            MATCH (to)
+            WHERE ID(to) = {to_id}
+            MERGE (from)-[r:{edge_type} {properties}]-{direction}(to)
+            RETURN id(r)
+        """
+        result = tx.run(cypher_template_filler(base_cypher,data))
+        return result.single()[0]
+
+    def merge_studies(self,studies:DataFrame):
         node_merging_func = self._merge_node
         with self._driver.session() as session:
-            logger.info("> Importing Studies Job is Started")
+            logger.info("> Merging Studies Job is Started")
             count_node = 0
             prev_count_node = 0
             
-            for study in studies.T.to_dict().values():
+            for study in studies.to_dict('records'):
                 node_type = "Study"
                 properties:dict = study
-                session.write_transaction(node_merging_func, node_type, properties)
+                study_id = session.write_transaction(node_merging_func, node_type, properties)
+                self.study_triald_and_neo4j_id_pairs[study["trial_id"]] = study_id
                 count_node += 1
-                if count_node > prev_count_node + 100:
+                if count_node > prev_count_node + 1000:
                     prev_count_node = count_node
-                    logger.info("> {} nodes already imported".format(count_node)) 
+                    logger.info("> {} nodes already merged".format(count_node)) 
```

Comment:
non-blocker, but generally better to do 
`logger.info("blah %s blah", myvar)` so when in production mode, string interpolation methods don't execute ( => faster), esp. good when in data loops like this

Reply:
do you want me to update it right now?

Reply:
Up to you. Fine merging as-is, I just like being helpful in code reviews as a mutual learning process :)